File: day-5/day-5.pl
#!/usr/bin/env perl
################################################################################
# Copyright (C) 2018 Michael Wiseman                                           #
#                                                                              #
# This program is free software: you can redistribute it and/or modify it      #
# under the terms of the GNU Affero General Public License as published by the #
# Free Software Foundation, either version 3 of the License, or (at your       #
# option) any later version.                                                   #
#                                                                              #
# This program is distributed in the hope that it will be useful, but WITHOUT  #
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        #
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License #
# for more details.                                                            #
#                                                                              #
# You should have received a copy of the GNU Affero General Public License     #
# along with this program.  If not, see <https://www.gnu.org/licenses/>.       #
################################################################################
use strict;
use warnings;
use List::MoreUtils 'all';
use feature 'say';

# Function definitions

sub part_one
{
    my $polymers = shift;
    my @letter_pairs;
    for my $ltr ('a'..'z')
    {
        push @letter_pairs, $ltr . (uc $ltr);
        push @letter_pairs, (uc $ltr) . $ltr;
    }
    my $found = 'true';
    while ($found eq 'true')
    {
        for my $pair (@letter_pairs)
        {
            $polymers =~ s/$pair//g;
        }
        if (all {index($polymers, $_) == -1} @letter_pairs)
        {
            $found = 'false';
        }
    }
    say length $polymers;
}

sub part_two
{
    my $polymers = shift;
    my @letter_pairs;
    my %letter_strprs;
    for my $ltr ('a'..'z')
    {
        push @letter_pairs, $ltr . (uc $ltr);
        push @letter_pairs, (uc $ltr) . $ltr;
        my $ucltr = uc $ltr;
        $letter_strprs{$ltr} = $polymers =~ s/$ltr|$ucltr//gr;
    }
    my $shortest_chain = 50000;
    while (my ($key, $value) = each %letter_strprs)
    {
        my $found = 'true';
        while ($found eq 'true')
        {
            for my $pair (@letter_pairs)
            {
                $value =~ s/$pair//g;
            }
            if (all {index($value, $_) == -1} @letter_pairs)
            {
                $found = 'false';
            }
        }
        $shortest_chain = length $value if $shortest_chain > length $value;
    }
    say $shortest_chain;
}

my $filename = 'day-5-input.txt';
open(my $ifh, '<', $filename);
my $input_file = <$ifh>;
close $ifh;
chomp($input_file);
# part_one($input_file); # 11364
part_two($input_file); # 4212

File: day-5/day-5.rb
#!/usr/bin/env ruby
################################################################################
# Copyright (C) 2018 Michael Wiseman                                           #
#                                                                              #
# This program is free software: you can redistribute it and/or modify it      #
# under the terms of the GNU Affero General Public License as published by the #
# Free Software Foundation, either version 3 of the License, or (at your       #
# option) any later version.                                                   #
#                                                                              #
# This program is distributed in the hope that it will be useful, but WITHOUT  #
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        #
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License #
# for more details.                                                            #
#                                                                              #
# You should have received a copy of the GNU Affero General Public License     #
# along with this program.  If not, see <https://www.gnu.org/licenses/>.       #
################################################################################

def part_one(input)
  polymers = input[0].chomp
  letter_pairs = []
  ('a'..'z').each do |ltr|
    letter_pairs.append(ltr + ltr.upcase)
    letter_pairs.append(ltr.upcase + ltr)
  end
  found = true
  loop do
    letter_pairs.each do |pair|
      polymers.gsub!(pair, '')
    end
    found = letter_pairs.any? {|pr| polymers.include? pr}
    break if found == false
  end
  puts polymers.length
end

def part_two(input)
  polymers = input[0].chomp
  letter_pairs = []
  letter_strprs = {}
  ('a'..'z').each do |ltr|
    letter_pairs.append(ltr + ltr.upcase)
    letter_pairs.append(ltr.upcase + ltr)
    letter_strprs[ltr] = polymers.gsub(/#{ltr}|#{ltr.upcase}/, '')
  end
  shortest_chain = 50000
  found = true
  letter_strprs.each do |key, value|
    loop do
      letter_pairs.each do |pair|
        value.gsub!(pair, '')
      end
      found = letter_pairs.any? {|pr| value.include? pr}
      break if found == false
    end
    shortest_chain = value.length if shortest_chain > value.length
  end
  puts shortest_chain
end

input_file = File.readlines('day-5-input.txt')
# part_one(input_file) # 11364
part_two(input_file) # 4212
